src/WebJobs.Script/Host/ScriptHostManager.cs

```diff
@@ -55,31 +68,60 @@ public void RunAndBlock(CancellationToken cancellationToken = default(Cancellati
                 // signaled. That is fine - the restart will be processed immediately
                 // once we get to this line again. The important thing is that these
                 // restarts are only happening on a single thread.
-                _instance.RestartEvent.WaitOne();
+                newInstance.RestartEvent.WaitOne();
 
-                // stop the host fully
-                _instance.Stop();
-                _instance.Dispose();
+                // Orphan the current host instance. We're stopping it, so it won't listen for nay new functions
+                // it will finish any currently executing functions and then clean itself up.
+                // Spin around and create a new host instance.
+                Task tIgnore = Orphan(newInstance);
             }
             while (!_stopped);
         }
 
+        // Let the existing host instance finsih currently executing functions.
+        private async Task Orphan(ScriptHost instance)
+        {
+            await instance.StopAsync();
```

Comment:
Please test this with a TimerTrigger. The issue there is that we need all Timer listeners to stop before we try to start the new host, since timers use a Singleton listener that is holding onto a blob lease - when the listener is stopped the lease is released (see [code here](https://github.com/Azure/azure-webjobs-sdk/blob/master/src/Microsoft.Azure.WebJobs.Host/Singleton/SingletonListener.cs#L77)). If that doesn't happen before the new host starts, the new host can't aquire the lease, and won't retry again for another minute.